auth_caching_sha2.go

```diff
@@ -0,0 +1,187 @@
+// Go MySQL Driver - A MySQL-Driver for Go's database/sql package
+//
+// Copyright 2023 The Go-MySQL-Driver Authors. All rights reserved.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package mysql
+
+import (
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+)
+
+// Authentication response constants
+const (
+	cachingSha2FastAuth       = 3 // Password found in cache
+	cachingSha2FullAuthNeeded = 4 // Full authentication needed
+)
+
+// CachingSha2PasswordPlugin implements the caching_sha2_password authentication
+// This plugin provides secure password-based authentication using SHA256 and RSA encryption,
+// with server-side caching of password verifiers for improved performance.
+type CachingSha2PasswordPlugin struct {
+	AuthPlugin
+}
+
+func init() {
+	RegisterAuthPlugin(&CachingSha2PasswordPlugin{})
+}
+
+func (p *CachingSha2PasswordPlugin) PluginName() string {
+	return "caching_sha2_password"
+}
+
+// InitAuth initializes the authentication process by scrambling the password.
+//
+// The scrambling process uses a three-step SHA256 hash:
+// 1. SHA256(password)
+// 2. SHA256(SHA256(password))
+// 3. XOR(SHA256(password), SHA256(SHA256(SHA256(password)), scramble))
+func (p *CachingSha2PasswordPlugin) InitAuth(authData []byte, cfg *Config) ([]byte, error) {
+	return scrambleSHA256Password(authData, cfg.Passwd), nil
+}
+
+// ContinuationAuth processes the server's response to our authentication attempt.
+//
+// The authentication flow can take several paths:
+//  1. Fast auth success (password found in cache)
+//  2. Full authentication needed:
+//     a. With TLS: send cleartext password
+//     b. Without TLS:
+//     - Request server's public key if not cached
+//     - Encrypt password with RSA public key
+//     - Send encrypted password
+func (p *CachingSha2PasswordPlugin) continuationAuth(packet []byte, authData []byte, mc *mysqlConn) ([]byte, error) {
+	if len(packet) == 0 {
+		return nil, fmt.Errorf("%w: empty auth response packet", ErrMalformPkt)
+	}
+
+	switch packet[0] {
+	case iOK, iERR, iEOF:
+		return packet, nil
+	case iAuthMoreData:
+		switch len(packet) {
+		case 1:
+			return mc.readPacket() // Auth successful
```

Comment:
i've keeped this part because was existing, but this is not in protocol https://dev.mysql.com/doc/dev/mysql-server/latest/page_caching_sha2_authentication_exchanges.html, server is expected to either send result fast_auth_success(3) or perform_full_authentication(4) only